interfaces/builtin/posix_mq.go

```diff
@@ -0,0 +1,237 @@
+		for _, perm := range perms {
+			if !iface.isValidPermission(perm) {
+				return nil, fmt.Errorf(`posix-mq slot %s permission "%s" not valid, must be one of %v`, name, perm, posixMQPlugPermissions)
+			}
+		}
+		return perms, nil
+	} else {
+		return posixMQDefaultPlugPermissions, nil
+	}
+}
+
+func (iface *posixMQInterface) getPath(attrs interfaces.Attrer, name string) (string, error) {
+	if pathAttr, isSet := attrs.Lookup("path"); isSet {
+		if path, ok := pathAttr.(string); ok {
+			return path, nil
+		}
+		return "", fmt.Errorf(`posix-mq slot %s "path" attribute must be a string, not %v`, name, pathAttr)
+	}
+	return "", fmt.Errorf(`posix-mq slot %s has missing "path" attribute`, name)
+}
+
+func (iface *posixMQInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	if path, err := iface.getPath(slot, slot.Name); err == nil {
+		if posixMQNamePattern.MatchString(path) {
+			return nil
+		}
+		return fmt.Errorf("posix-mq path must conform to the POSIX message queue name specifications (see `man mq_overview`)")
+	} else {
+		return err
+	}
+}
+
+func (iface *posixMQInterface) AutoConnect(plug *snap.PlugInfo, slot *snap.SlotInfo) bool {
+	return true
+}
+
+func (iface *posixMQInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	if !implicitSystemPermanentSlot(slot) {
+		if path, err := iface.getPath(slot, slot.Name); err == nil {
+			spec.AddSnippet(fmt.Sprintf(`# POSIX Message Queue management
+mqueue (create delete getattr setattr read write) %s,
+`, path))
+		} else {
+			return err
+		}
+	}
+	return nil
+}
+
+func (iface *posixMQInterface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	if path, err := iface.getPath(slot, slot.Name()); err == nil {
+		spec.AddSnippet(fmt.Sprintf(`# POSIX Message Queue slot communication
+mqueue (create delete getattr setattr) %s,
+`, path))
+	} else {
+		return err
+	}
+	return nil
+}
+
+func (iface *posixMQInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	if path, err := iface.getPath(slot, slot.Name()); err == nil {
+		if perms, err := iface.getPermissions(slot, slot.Name()); err == nil {
+			aaPerms := strings.Join(perms, " ")
+			spec.AddSnippet(fmt.Sprintf(`# POSIX Message Queue plug communication
+mqueue (%s) %s,
+`, aaPerms, path))
+		} else {
+			return err
+		}
+	} else {
+		return err
+	}
+	return nil
+}
+
+func (iface *posixMQInterface) SecCompPermanentSlot(spec *seccomp.Specification, slot *snap.SlotInfo) error {
+	spec.AddSnippet(posixMQPermanentSlotSecComp)
+	return nil
+}
+
+func (iface *posixMQInterface) SecCompConnectedPlug(spec *seccomp.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	if perms, err := iface.getPermissions(slot, slot.Name()); err == nil {
+		// always allow these functions
+		var syscalls = []string{
+			"mq_notify",
```

Comment:
Should `mq_notify` go with `mq_read`? ie. we should only allow `mq_notify` if we also allow `mq_read` since both allow to get information about what is in the queue.